### DISCUSSION TITLE
Broke my LCD module, any chance this would work as a replacement?

### DISCUSSION THREAD

**Discussion by filmgeek47** [sanni/cartreader#1057] [Need help]
Broke my LCD module. while troubleshooting an error code. I know the wiki specifically recommends against buying non mks branded modules, but to my eyes this one looks identical:
https://www.amazon.com/dp/B0CLM5GW7T?ref=ppx_yo2ov_dt_b_fed_asin_title

What are the odds this would work? I'm mid build and the other options won't come for weeks.

**Comment by gemarcano** [comment]
What broke about your LCD module? If the SD card slot and power all still work, and the issue is with the screen or buttons, you can probably hobble along with SERIAL_MONITOR in Config.h, and use the Arduino IDE's serial monitor interface to control the hardware.

The Amazon listing looks like mine, which I got from AliExpress, except it's missing some branding. If your SD card hardware works, I'd just order from AliExpress and wait until that comes, and if you're willing to possibly burn $20 I'd get the Amazon one and check if it works.

**Comment by filmgeek47** [comment]
To be honest I’m not sure what I broke. There’s a tiny black surface mount component next to the SD card slot on the back of the module that I snapped (freak accident). No idea what it drives. Prior to that I was getting an “SD error” when powering on, so there’s a chance the card slot was bad anyway (I tried two SD cards and got the same thing both times.

Might just burn the $20. Thanks for your help.

**Reply by gemarcano** [reply]
That might be a resistor, R14, if it's right behind the SD card adapter. Markings on mine are `01C` which corresponds to a 10kOhm resistor. Looks to be an imperial 0603 package size. This may well be just a pull-up or pull-down resistor. If you're  feeling brave and have a soldering iron somewhere, you can order 0603 SMD 10kOhm resistors from Digikey/Mouser or even from AliExpress for fairly cheap (you can probably get 1000 and shipping may even still be more than the stupid components, lol). 10kOhm resistor is also fairly common in general... so it's likely any scrap electronics you may have might even have one (now, would they be the right size?).

**Reply by gemarcano** [reply]
I found the schematics for the modules. Yes, R14 is a pull-up resistor, and it's used to provide the SD card detection pin. When an SD card is plugged in, that whole line is pulled down to ground. When it isn't, the pull-up pulls the voltage up to 5V. Honestly, if it's R14 that's broken, I think your device should work with the SD card plugged in. That line should be pulled down by the SD card port even without the resistor. The hardware just may not be able to detect the SD card missing.

**Reply by filmgeek47** [reply]
Thanks so much for looking into it. Yeah that's definitely the part I broke. I hooked everything back together and nothing exploded so assuming you're right and it's effectively an unnecessary component in this case, I'm no worse off than I was before. Unfortunately that still leaves me trying to work out why I was getting an SD Error in the first place. I've tried 3 SD cards now (two 32GB microSD, one 64GB full size SD because it's what I had handy). 

I'm pretty confident I haven't bridged anything on the parts I've soldered. Checked adjacent parts with my continuity meter and didn't find anything amiss. Not sure what I'm missing here.

**Reply by gemarcano** [reply]
The part is necessary-ish (to help detect that the SD card isn't inserted), but for testing you might be able to do away with it for the time being.

I did run into weird issues when I was first trying to get the system running with an SD card. I ended up finding this discussion which was useful for debugging: https://github.com/sanni/cartreader/discussions/686#discussioncomment-4754195

That particular reply has a picture of which pins are the ones that handle SD card comms, so I'd check continuity there first (one of the times I inserted my Mega the pins managed to miss the holes apparently, that's how I caught this). The rest of the issue discussion is useful for trouble shooting.

Of note, if you enable `SERIAL_MONITOR` on your `Config.h` and upload, do you see `SD Error` in the serial monitor in the Arduino IDE? That mode pretty much doesn't use most of the display module, just the SD card, so it's easier to try to track down what's going on.

I really wish the SD card library in use would give better error conditions (it literally just returns `true` on success, and `false` on failures).

**Reply by Ancyker** [reply]
SPI_DET (SD_CD) is NC on HW5:
![image](https://github.com/user-attachments/assets/1cfc3855-9b5e-4dd6-9603-5f2bccaca1b7)

**Reply by filmgeek47** [reply]
Okay so I've done some more digging. I actually ordered the amazon listing, and it turned out to be an identical replacement to the module I broke. Unfortunately that didn't fix my SD card error, so it's not a bad reader. 

I then started resoldering pins D50-53 on both the Arduino and the main sanni PCB, which also didn't fix it. HOWEVER, pulling out my multimeter, I discovered I'm reading continuity betwee D50, and 51. I'm assuming that's not normal, but for the life of me I can't find the bridge. I desoldered the pins on both the female Arduino header side and the outputs to the LCD side, then rechecked with a magnifier and they're not touching. Also desoldered the single header row. I'm really confused. 

Any ideas?

**Reply by gemarcano** [reply]
I've heard the soldering of the atmega chip itself on the Arduino can be... suspect. I'd pull the Arduino out of the cart reader and check for shorts between the pins corresponding to D50 and D51 on the isolated Arduino. If there's a short, inspect the big chip on the Arduino for bridged pins. Apparently D50 and D51 are MOSI and SCK, and are pins 20 and 21 on the atmega chip (Arduino schematic that may help: https://redmine.laas.fr/attachments/download/3867/MEGA-PRO-CH340GATmega2560.pdf). Those two pins are next to each other so there's a non zero chance of a bridge.

**Reply by filmgeek47** [reply]
Thanks. So strangely the short isn’t on the arduino. I’m getting them on the main PCB when the arduino’s removed.

**Reply by gemarcano** [reply]
That's bizarre. Here are pictures of the HW5 PCB layout for D50 and D51. The main areas where there might be shorts are by the Arduino headers on the mainboard, or near the GBA port in the SPI area. It's possible there's a short under the EXP2 IDC socket for the screen, but D50 and D51 aren't as close to each other there.

Well, the next steps depend on how you want to approach the issue. A visual inspection of the sites I mentioned might help. Additionally, a more destructive approach would be to start cutting the D50 and D51 traces strategically to see if you can figure out where the short is coming from. A small cut is simple enough to repair with a solder bridge.

Blindly, I would add flux and reflow the SPI pins of the female pin header by the GBA port-- if I were a betting person I suspect a short could be underneath the header from possibly too much solder (maybe). It's unlikely, honestly, but hey, obviously something is shorting.

![image](https://github.com/user-attachments/assets/031535ca-820f-495d-b223-3a255aeb8bd3)
![image](https://github.com/user-attachments/assets/8d42112c-51d5-40a6-85e9-f499959b26a5)

**Reply by filmgeek47** [reply]
Finally solved it. Remy pointed out that there’s a spot where you can get away with cutting the D50 and D51 traces without compromising functionality, as the initial section of the traces is basically just there from prototyping. Never found the short, but cutting those traces did the trick and the reader’s now working great on all the cartridge slots. Now I just need to reassemble it inside that damn shell. Aligning those pins blind is no easy feat. 😂 Thanks again for your help.

**Comment by filmgeek47** [comment]
This is really helpful, thanks. I’ll dive into this and see what I can figure out.
